lib/matplotlib/axes/_axes.py

```diff
@@ -4014,14 +4014,15 @@ def scatter(self, x, y, s=None, c=None, marker=None, cmap=None, norm=None,
         else:
             try:
                 c_array = np.asanyarray(c, dtype=float)
-                if c_array.size == x.size:
```

Comment:
Looking back at @anntzer's comment in #7314, I think this line should be checking that c_array.shape matches the *original* shape of x and y.  That implies that x and y should be checked earlier for matching shapes, and that their original common shape should be saved so it can be used in the c shape-check.